File: src/mcp/services/skills.py
"""
MCP Skills Service

Skills service FastMCP module with 4 tools for skill operations.
"""

import logging
from mcp.server.fastmcp import FastMCP

logger = logging.getLogger("niko-gateway")

skills_mcp = FastMCP("NikoSkills", stateless_http=True)


@skills_mcp.tool()
async def skills_list(category: str = None) -> list:
    """
    列出所有技能包

    Args:
        category: 类别过滤 (character/structure/suspense/description/evaluation)

    Returns:
        技能列表
    """
    from src.agents.skill_router import SkillRouter

    router = SkillRouter()
    all_skills = router.list_all_skills()

    result = []
    for skill_id, info in all_skills.items():
        if category and info.get("category", "") != category:
            continue
        skill_entry = {
            "id": skill_id,
            "name": info.get("name", skill_id),
            "description": info.get("description", ""),
            "keywords": info.get("keywords", [])
        }
        result.append(skill_entry)

    return result


@skills_mcp.tool()
async def skills_match(
    task_type: str = None,
    keywords: list = None,
    issue: str = None
) -> list:
    """
    匹配技能包

    Args:
        task_type: 任务类型 (character_creation/dialogue/climax_writing 等)
        keywords: 关键词列表
        issue: 问题描述

    Returns:
        匹配的技能列表，按相关度排序
    """
    from src.agents.skill_router import SkillRouter, TaskType

    router = SkillRouter()

    if task_type:
        try:
            tt = TaskType(task_type)
            recommendations = router.route_by_task_type(tt)
        except ValueError:
            recommendations = []
    elif keywords:
        recommendations = router.route_by_keywords(keywords)
    elif issue:
        recommendations = router.route_by_issue(issue)
    else:
        recommendations = []

    return [
        {
            "skill_id": r.skill_id,
            "skill_name": r.skill_name,
            "relevance": r.relevance,
            "reason": r.reason,
            "priority": r.priority
        }
        for r in recommendations
    ]


@skills_mcp.tool()
async def skills_load(skill_id: str) -> dict:
    """
    加载技能包内容

    Args:
        skill_id: 技能ID

    Returns:
        技能包完整内容
    """
    from src.skills.skill_loader import SkillLoader

    loader = SkillLoader()
    skill = loader.load_skill(skill_id)

    if skill:
        return {
            "id": skill_id,
            "content": skill.get("content", ""),
            "metadata": skill.get("metadata", {})
        }

    return {"error": f"Skill '{skill_id}' not found"}


@skills_mcp.tool()
async def skills_get_chain(task_type: str) -> list:
    """
    获取技能链 (按执行顺序)

    Args:
        task_type: 任务类型

    Returns:
        按执行顺序排列的技能链
    """
    from src.agents.skill_router import SkillRouter, TaskType

    router = SkillRouter()

    try:
        tt = TaskType(task_type)
        chain = router.get_skill_chain(tt)
        return [
            {
                "skill_id": r.skill_id,
                "skill_name": r.skill_name,
                "step": r.priority,
                "reason": r.reason
            }
            for r in chain
        ]
    except ValueError:
        return []


__all__ = ["skills_mcp"]

File: src-ts/mcp/services/skills.ts
/**
 * MCP Skills Service
 *
 * Skills service module with 4 tools for skill operations.
 * Ported from src/mcp/services/skills.py
 */

import {
  SkillRouter,
  TaskType,
  type SkillRecommendation as RouterSkillRecommendation,
} from '../../agents/skill-router';
import { SkillLoader } from '../../skills/skill-loader';

// ---------------------------------------------------------------
// Skill recommendation type
// ---------------------------------------------------------------

export interface SkillRecommendation {
  skill_id: string;
  skill_name: string;
  relevance: number;
  reason: string;
  priority: number;
}

type SkillListItem = { id: string; name: string; description: string; keywords: string[] };

let skillRouterInstance: SkillRouter | null = null;
let skillLoaderInstance: SkillLoader | null = null;

function getSkillRouter(): SkillRouter {
  if (!skillRouterInstance) {
    skillRouterInstance = new SkillRouter();
  }
  return skillRouterInstance;
}

function getSkillLoader(): SkillLoader {
  if (!skillLoaderInstance) {
    skillLoaderInstance = new SkillLoader();
  }
  return skillLoaderInstance;
}

function mapRecommendation(rec: RouterSkillRecommendation): SkillRecommendation {
  return {
    skill_id: rec.skillId,
    skill_name: rec.skillName,
    relevance: rec.relevance,
    reason: rec.reason,
    priority: rec.priority,
  };
}

function normalizeTaskType(taskType?: string | null): TaskType | null {
  if (!taskType) return null;
  const normalized = taskType.trim().toLowerCase();
  const values = Object.values(TaskType) as string[];
  return values.includes(normalized) ? (normalized as TaskType) : null;
}

function dedupeRecommendations(items: SkillRecommendation[]): SkillRecommendation[] {
  const byId = new Map<string, SkillRecommendation>();

  for (const item of items) {
    const existing = byId.get(item.skill_id);
    if (
      !existing ||
      item.relevance > existing.relevance ||
      (item.relevance === existing.relevance && item.priority < existing.priority)
    ) {
      byId.set(item.skill_id, item);
    }
  }

  return [...byId.values()].sort((left, right) =>
    right.relevance - left.relevance || left.priority - right.priority || left.skill_id.localeCompare(right.skill_id),
  );
}

// ---------------------------------------------------------------
// Tool implementations
// ---------------------------------------------------------------

export async function skillsList(
  category?: string | null
): Promise<SkillListItem[]> {
  const loader = getSkillLoader();
  const summaries = loader.getSummaryDict() as Array<Record<string, unknown>>;
  const normalizedCategory = (category ?? '').trim().toLowerCase();

  const items = summaries.map((item) => ({
    id: String(item.name ?? ''),
    name: String(item.name ?? ''),
    description: String(item.description ?? ''),
    keywords: [
      ...(Array.isArray(item.tags) ? item.tags : []).map((tag) => String(tag)),
      ...(Array.isArray(item.triggers) ? item.triggers : []).map((trigger) => String(trigger)),
    ],
  }));

  if (!normalizedCategory) {
    return items;
  }

  return items.filter((item) =>
    item.id.toLowerCase().includes(normalizedCategory) ||
    item.name.toLowerCase().includes(normalizedCategory) ||
    item.description.toLowerCase().includes(normalizedCategory) ||
    item.keywords.some((keyword) => keyword.toLowerCase().includes(normalizedCategory)),
  );
}

export async function skillsMatch(params: {
  taskType?: string | null;
  keywords?: string[] | null;
  issue?: string | null;
}): Promise<SkillRecommendation[]> {
  const router = getSkillRouter();
  const recommendations: SkillRecommendation[] = [];
  const taskType = normalizeTaskType(params.taskType);

  if (taskType) {
    recommendations.push(...router.routeByTaskType(taskType).map(mapRecommendation));
  }

  if (Array.isArray(params.keywords) && params.keywords.length > 0) {
    recommendations.push(...router.routeByKeywords(params.keywords).map(mapRecommendation));
  }

  if (typeof params.issue === 'string' && params.issue.trim()) {
    recommendations.push(...router.routeByIssue(params.issue.trim()).map(mapRecommendation));
  }

  return dedupeRecommendations(recommendations);
}

export async function skillsLoad(
  skillId: string
): Promise<{ id: string; content: string; metadata: Record<string, unknown> } | { error: string }> {
  const normalizedSkillId = skillId.trim();
  if (!normalizedSkillId) {
    return { error: "Skill '' not found" };
  }

  try {
    const loader = getSkillLoader();
    const loaded = loader.loadSkill(normalizedSkillId) as Record<string, unknown>;
    return {
      id: normalizedSkillId,
      content: String(loaded.content ?? ''),
      metadata: (loaded.metadata as Record<string, unknown>) ?? {},
    };
  } catch {
    return { error: `Skill '${normalizedSkillId}' not found` };
  }
}

export async function skillsGetChain(
  taskType: string
): Promise<Array<{ skill_id: string; skill_name: string; step: number; reason: string }>> {
  const normalizedTaskType = normalizeTaskType(taskType);
  if (!normalizedTaskType) {
    return [];
  }

  const router = getSkillRouter();
  return router.getSkillChain(normalizedTaskType).map((item, index) => ({
    skill_id: item.skillId,
    skill_name: item.skillName,
    step: index + 1,
    reason: item.reason,
  }));
}
